src/storage.rs

```diff
@@ -0,0 +1,198 @@
+/// The `scope_hash` represents the signature of the scopes for which the given token
+/// should be stored or retrieved.
+/// For completeness, the underlying, sorted scopes are provided as well. They might be
+/// useful for presentation to the user.
+pub trait TokenStorage {
+    type Error: 'static + Error;
+
+    /// If `token` is None, it is invalid or revoked and should be removed from storage.
+    /// Otherwise, it should be saved.
+    fn set(&mut self,
+           scope_hash: u64,
+           scopes: &Vec<&str>,
+           token: Option<Token>)
+           -> Result<(), Self::Error>;
+    /// A `None` result indicates that there is no token for the given scope_hash.
+    fn get(&self, scope_hash: u64, scopes: &Vec<&str>) -> Result<Option<Token>, Self::Error>;
+}
+
+/// A storage that remembers nothing.
+#[derive(Default)]
+pub struct NullStorage;
+#[derive(Debug)]
+pub struct NullError;
+
+impl Error for NullError {
+    fn description(&self) -> &str {
+        "NULL"
+    }
+}
+
+impl fmt::Display for NullError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        "NULL-ERROR".fmt(f)
+    }
+}
+
+impl TokenStorage for NullStorage {
+    type Error = NullError;
+    fn set(&mut self, _: u64, _: &Vec<&str>, _: Option<Token>) -> Result<(), NullError> {
+        Ok(())
+    }
+    fn get(&self, _: u64, _: &Vec<&str>) -> Result<Option<Token>, NullError> {
+        Ok(None)
+    }
+}
+
+/// A storage that remembers values for one session only.
+#[derive(Default)]
+pub struct MemoryStorage {
+    pub tokens: HashMap<u64, Token>,
+}
+
+impl TokenStorage for MemoryStorage {
+    type Error = NullError;
+
+    fn set(&mut self,
+           scope_hash: u64,
+           _: &Vec<&str>,
+           token: Option<Token>)
+           -> Result<(), NullError> {
+        match token {
+            Some(t) => self.tokens.insert(scope_hash, t),
+            None => self.tokens.remove(&scope_hash),
+        };
+        Ok(())
+    }
+
+    fn get(&self, scope_hash: u64, _: &Vec<&str>) -> Result<Option<Token>, NullError> {
+        match self.tokens.get(&scope_hash) {
+            Some(t) => Ok(Some(t.clone())),
+            None => Ok(None),
+        }
+    }
+}
+
+/// A single stored token.
+#[derive(Serialize, Deserialize)]
+struct JSONToken {
+    pub hash: u64,
+    pub token: Token,
+}
+
+/// List of tokens in a JSON object
+#[derive(Serialize, Deserialize)]
+struct JSONTokens {
+    pub tokens: Vec<JSONToken>,
+}
+
+/// Serializes tokens to a JSON file on disk.
+#[derive(Default)]
+pub struct DiskTokenStorage {
+    location: String,
+    tokens: HashMap<u64, Token>,
+}
+
+impl DiskTokenStorage {
+    pub fn new(location: &String) -> DiskTokenStorage {
+        use std::fs;
+        let mut dts = DiskTokenStorage {
+            location: location.clone(),
+            tokens: HashMap::new(),
+        };
+
+        // best-effort
+        let _ = dts.load_from_file();
```

Comment:
Ok, I'll just make `new()` return a `Result` type.

My thoughts were that `new()` is just "make me a storage; if you can't read it, ignore it, I'll find out when there's no token in it". But you're absolutely right with regards to The Right Way To Do It :-)